File: 034_Find_First_and_Last_Position_of_Element_in_Sorted_Array/034_1.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> res(2, -1);
        if (nums.empty()) return res;
        if (nums.size() == 1) {
            if (target == nums[0]) return {0, 0};
            else                   return {-1, -1};
        }
        int lo = 0, hi = nums.size();
        while (lo < hi) {
            int mid = (lo + hi) >> 1;
            if (target <= nums[mid])
                hi = mid;
            else 
                lo = mid + 1;
        }
        if (lo == nums.size() || nums[lo] != target) return res;
        res[0] = lo;

        hi = nums.size();
        while (lo < hi) {
            int mid = (lo + hi) >> 1;
            if (target >= nums[mid])
                lo = mid + 1;
            else
                hi = mid;
        }
        res[1] = hi - 1;
        return res;
    }
};

File: 034_Find_First_and_Last_Position_of_Element_in_Sorted_Array/034_1.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int start = search(nums, target);
        int end = search(nums, target + 1);
        if (start == nums.length || nums[start] != target)   // prevent start = nums.length, array out of bounds
            return new int[]{-1, -1};
        return new int[]{start, end - 1};
    }

    private int search(int[] nums, int target) {
        int lo = 0, hi = nums.length;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (target <= nums[mid])
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo;
    }
}

public class leetcode {
    public static void main(String[] args) {
        int[] nums = {5, 7, 7, 8, 8, 10};
        Solution s = new Solution();
        System.out.println(s.searchRange(nums, 8)[0]);
    }
}